### PR TITLE
nixos/grafana-agent: update example to merge with defaults

### PR DESCRIPTION
**Pull request by flokli**
It seems defaults are not merged, so `lib.mkMerge` needs to be used.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zimbatm** [comment]
What if the default was set in the config, would it make it mergable?

**Comment by flokli** [comment]
> What if the default was set in the config, would it make it mergable?

I'm not sure I understand the question, sorry.

**Comment by zimbatm** [comment]
I think for most people, the assumption would be that new config values would get merged into the default. And they would have to use something like mkForce to explicitly replace the default value.

How can we change the option so it would work that way?

**Comment by flokli** [comment]
Summoning @roberth and @infinisil ;-)

File: nixos/modules/services/monitoring/grafana-agent.nix [hunk 1]
```diff
@@ -61,45 +61,50 @@ in
         };
```
Comment (Mic92) [inline nixos/modules/services/monitoring/grafana-agent.nix:61]: I think to have them merged, you can move the defaults to `config` section of this module and have them set using `lib.mkDefault`. Than they should get merged with the users supplied configuration.

**Comment by roberth** [comment]
> Summoning @roberth and @infinisil ;-)

A @roberth materializes and agrees with @zimbatm and @Mic92. He adds that the `mkDefault` should be applied to the individual values, because it's the _whole_ `mkDefault` argument that gets replaced.

**Comment by flokli** [comment]
@roberth @Mic92 like this?

```patch
From 04a39ae35854f0daa5f22d6d2808e9ac3ba8a7f5 Mon Sep 17 00:00:00 2001
From: Florian Klink <flokli@flokli.de>
Date: Thu, 28 Jul 2022 13:42:16 +0700
Subject: [PATCH] nixos/grafana-agent: ensure defaults are merged

Move the defaults to the `config` section of the module, and apply them
with mkDefault.

That way the defaults are merged with user-provided config, and are
merged without having to use lib.mkForce.
---
 .../services/monitoring/grafana-agent.nix     | 27 ++++++++++---------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/nixos/modules/services/monitoring/grafana-agent.nix b/nixos/modules/services/monitoring/grafana-agent.nix
index bbeda184647..bf9511fa92e 100644
--- a/nixos/modules/services/monitoring/grafana-agent.nix
+++ b/nixos/modules/services/monitoring/grafana-agent.nix
@@ -48,19 +48,7 @@ in
         freeformType = settingsFormat.type;
       };
 
-      default = {
-        metrics = {
-          wal_directory = "\${STATE_DIRECTORY}";
-          global.scrape_interval = "5s";
-        };
-        integrations = {
-          agent.enabled = true;
-          agent.scrape_integration = true;
-          node_exporter.enabled = true;
-          replace_instance_label = true;
-        };
-      };
-
+      default = { };
       example = {
         metrics.global.remote_write = [{
           url = "\${METRICS_REMOTE_WRITE_URL}";
@@ -104,6 +92,19 @@ in
   };
 
   config = mkIf cfg.enable {
+    services.grafana-agent.settings = {
+      metrics = {
+        wal_directory = mkDefault "\${STATE_DIRECTORY}";
+        global.scrape_interval = mkDefault "5s";
+      };
+      integrations = {
+        agent.enabled = mkDefault true;
+        agent.scrape_integration = mkDefault true;
+        node_exporter.enabled = mkDefault true;
+        replace_instance_label = mkDefault true;
+      };
+    };
+
     systemd.services.grafana-agent = {
       wantedBy = [ "multi-user.target" ];
       script = ''
-- 
2.36.1

```

**Comment by roberth** [comment]
lgtm

**Comment by flokli** [comment]
PR up at https://github.com/NixOS/nixpkgs/pull/183269.

**Pull request closed without merging**